util/hot_backup.cc

```diff
@@ -0,0 +1,654 @@
+        //  (includes code that creates hard links to .sst files)
+        good=m_DBImpl.WriteBackupManifest();
+
+        // last step is to replicate as much of LOG file from time of
+        //  backup (close, but not exact)
+        //  Assumes that 0 for log_position implies user has a custom
+        //  Logger class that this routine will fail to copy.
+        if (good && 0!=log_position)
+            good=m_DBImpl.CopyLOGSegment(log_position);
+    }   // if
+
+    // inform db and master object that this db is done.
+    m_DBImpl.HotBackupComplete();
+    gHotBackup->HotBackupFinished();
+
+    return;
+
+}   // HotBackupTask::operator()
+
+
+/**
+ * Rotate directory names in a backup, backup.1, backup.2, etc. pattern.
+ *  This is similar to many /var/log files on unix systems.
+ */
+bool
+HotBackup::PrepareDirectories(
+    const Options & LiveOptions)
+{
+    Options local_options;
+    bool good;
+    Status status;
+
+    assert(0 < config::kNumBackups);
+    good=true;
+
+    // Destroy highest possible directory if it exists
+    local_options=LiveOptions;
+    good=SetBackupPaths(local_options, config::kNumBackups-1);
+    status=DestroyDB("", local_options);
+
+    // Rotate names of directories
+    if (status.ok())
+    {
+        int loop;
+
+        for (loop=config::kNumBackups-1; 1 <= loop && status.ok(); --loop)
+        {
+            int src, dest;
+            std::string src_path, dest_path;
+
+            src=loop-1;
+            dest=loop;
+
+            // rename fast / default tier
+            src_path=BackupPath(LiveOptions.tiered_fast_prefix, src);
+            dest_path=BackupPath(LiveOptions.tiered_fast_prefix, dest);
+
+            if (LiveOptions.env->FileExists(src_path))
+                status=LiveOptions.env->RenameFile(src_path, dest_path);
+
+            // rename slow tier
+            if (status.ok())
+            {
+                if (0<LiveOptions.tiered_slow_level)
+                {
+                    src_path=BackupPath(LiveOptions.tiered_slow_prefix, src);
+                    dest_path=BackupPath(LiveOptions.tiered_slow_prefix, dest);
+
+                    if (LiveOptions.env->FileExists(src_path))
+                        status=LiveOptions.env->RenameFile(src_path, dest_path);
+
+                    if (!status.ok())
+                    {
+                        good=false;
+                        Log(LiveOptions.info_log, "HotBackup failed while renaming %s slow directory",
+                            src_path.c_str());
+                    }   // if
+                }   // if
+            }
+            else
+            {
+                good=false;
+                Log(LiveOptions.info_log, "HotBackup failed while renaming %s directory",
+                    src_path.c_str());
+            }   // else
+        }   // for
+
+    }   // if
+    else
+    {
+        good=false;
+        Log(LiveOptions.info_log, "HotBackup failed while removing %s directory",
```

Comment:
This message will be misleading if the removal failed on the slow tier. Does the status object have any more info that could be logged?